pkg/ebpf/tracer_windows.go

```diff
@@ -115,22 +122,41 @@ func printStats(stats []network.ConnectionStats) {
 
 // GetActiveConnections returns all active connections
 func (t *Tracer) GetActiveConnections(clientID string) (*network.Connections, error) {
-	connStatsActive, connStatsClosed, err := t.driverInterface.GetConnectionStats()
+	t.bufferLock.Lock()
+	defer t.bufferLock.Unlock()
+
+	activeConnStats, closedConnStats, err := t.driverInterface.GetConnectionStats(t.connStatsActive[:0], t.connStatsClosed[:0])
 	if err != nil {
 		log.Errorf("failed to get connnections")
 		return nil, err
 	}
 
-	for _, connStat := range connStatsClosed {
+	for _, connStat := range closedConnStats {
 		t.state.StoreClosedConnection(connStat)
 	}
 
 	// check for expired clients in the state
 	t.state.RemoveExpiredClients(time.Now())
-	conns := t.state.Connections(clientID, uint64(time.Now().Nanosecond()), connStatsActive, t.reverseDNS.GetDNSStats())
+	conns := t.state.Connections(clientID, uint64(time.Now().Nanosecond()), activeConnStats, t.reverseDNS.GetDNSStats())
+	t.checkBufferResize(len(activeConnStats), len(closedConnStats))
 	return &network.Connections{Conns: conns}, nil
 }
 
+func (t *Tracer) checkBufferResize(activeConnSize int, closedConnSize int) {
+	// Grow or shrink buffer depending on the usage
+	if activeConnSize >= cap(t.connStatsActive)*2 {
+		t.connStatsActive = make([]network.ConnectionStats, 0, cap(t.connStatsActive)*2)
+	} else if activeConnSize <= cap(t.connStatsActive)/2 {
+		t.connStatsActive = make([]network.ConnectionStats, 0, cap(t.connStatsActive)/2)
+	}
+
+	if closedConnSize >= cap(t.connStatsClosed)*2 {
+		t.connStatsClosed = make([]network.ConnectionStats, 0, cap(t.connStatsClosed)*2)
+	} else if closedConnSize <= cap(t.connStatsClosed)/2 {
+		t.connStatsClosed = make([]network.ConnectionStats, 0, cap(t.connStatsClosed)/2)
```

Comment:
I think you want the max of `closedConnSize` and `cap(t.connStatsClosed)/2`, otherwise there is a great chance of `DriverInterface.GetConnectionStats` needing to grow the underlying array.

Side note: why shrink the buffer? If you are concerned about performance, you are doing an additional unnecessary allocation.

Reply:
Makes sense @brycekahle. The idea for shrinking the buffer is so that we aren't allocating space in a buffer we don't need. 

Think of a situation in which a large burst of connections that causes the buffer to grow, but in general we don't need a buffer of that size.